### ISSUE TITLE
[css-tables][css-sizing] What sizing keywords allow fixed table mode?

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#10937]
### Situation

https://drafts.csswg.org/css-tables/#table-layout-property says

> A table-root is said to be laid out in fixed mode whenever the computed value of the [`table-layout`](https://drafts.csswg.org/css-tables/#propdef-table-layout) property is equal to `fixed`, and the specified width of the table root is **either a `<length-percentage>`, `min-content` or `fit-content`**.

This is the logic that browsers use for the inline size:

 - WebKit: https://searchfox.org/wubkat/rev/b36cbce69fddb7da33823f316bd8ead5bebee970/Source/WebCore/rendering/style/RenderStyleInlines.h#526
   `logicalWidth().isSpecified() || logicalWidth().isFitContent() || logicalWidth().isMinContent()`
 - Firefox: https://searchfox.org/mozilla-central/rev/1e8cec3727d6e09f4af41bb3d202b7a4c326ed84/layout/tables/nsTableFrame.cpp#3482
   (negated) `iSize.IsAuto() || iSize.IsMaxContent()`
 - Chrome: https://source.chromium.org/chromium/chromium/src/+/main:third_party/blink/renderer/core/style/computed_style.h;drc=572789ee61c7e857aa288a7ac7768b2b0b4f4dcc;l=1924
   `!LogicalWidth().HasAuto() && !LogicalWidth().HasMaxContent()`

So we have a mismatch of keywords that accept fixed table layout:

| Value | Spec | WebKit | Gecko | Blink |
| - | - | - | - | - |
| `<length-percentage>`    | ✅ | ✅ | ✅ | ✅ |
| `auto`                   | ❌ | ❌ | ❌ | ❌ |
| `min-content` (*)        | ✅ | ✅ | ✅ | ✅ |
| `max-content` (*)        | ❌ | ❌ | ❌ | ❌ |
| `fit-content` (*)        | ✅ | ✅ | ✅ | ✅ |
| `<fit-content()>`        | ❌ | ❌︎ | ✅ | ✅︎ |
| `contain`                | ❌ | ❌︎ | ✅︎ | ✅︎ |
| `stretch`                | ❌ | ❌︎ | ✅ | ✅︎ |
| `-webkit-fill-available` | ❌︎ | ❌ | ✅ | ✅ |
| `-moz-available`         | ❌︎ | ❌︎ | ✅ | ✅︎ |
| `intrinsic`              | ❌︎ | ❌ | ✅︎ | ✅︎ |
| `min-intrinsic`          | ❌︎ | ❌ | ✅︎ | ✅︎ |

`<calc-size()>` should just behave as its basis.

(*) including `-webkit-` and `-moz-` prefixes.
✅: Supports the value (at least partially), and allows fixed table layout
✅︎: Doesn't support the value, but it would allow fixed table layout
❌: Supports the value (at least partially), and doesn't allows fixed table layout
❌︎: Doesn't support the value, but it wouldn't allow fixed table layout

### Example

```html
<!DOCTYPE html>
<div style="width: 0">
  <table style="table-layout: fixed; width: -webkit-fill-available" border>
    <tr><td>lorem ipsum</td></tr>
  </table>
</div>
```

On Firefox (with `layout.css.webkit-fill-available.enabled = true`) and Blink the table ignores the contents and is almost 0px wide, while on WebKit it's wider according to the text.

### Changes

I think that `<fit-content()>` should allow fixed table layout, just like `fit-content` and `min-content`.

And maybe `stretch` should too, to match `100%`? Also, `-webkit-fill-available`/`-moz-available` is kinda the same thing, and it allows fixed table layout on 2/3 of the major browsers.

Nobody implements `contain`, but we should also decide what to do with it.

**Comment by bfgeek** [comment]
As an aside does anyone know what max-content doesn't apply?

(FWIW the "rule" that we've been applying to to a lot of these table behaviours, is if "width is auto" do the special behaviour - this comes up a suprising amount in table layout).

**Comment by Loirooriol** [comment]
Yeah, it's kinda weird that, given enough available space, `fit-content` doesn't behave like `max-content`:

```html
<!DOCTYPE html>
<div style="width: max-content">
  <table style="table-layout: fixed; width: max-content" border>
    <tr><td style="width: 0">lorem ipsum</td></tr>
  </table>
  <table style="table-layout: fixed; width: fit-content" border>
    <tr><td style="width: 0">lorem ipsum</td></tr>
  </table>
</div>
```

Only checking for `auto` would make this more future-proof.

**Comment by Loirooriol** [comment]
There was a broken test in https://github.com/web-platform-tests/wpt/blob/master/css/css-tables/fixed-layout-2.html, I have fixed it and tested all the cases above to help with interoperability.

**Comment by Loirooriol** [comment]
https://github.com/WebKit/WebKit/pull/34706 aligned `-webkit-fill-available` with Blink and Gecko.

**Comment by tabatkins** [comment]
If `fit-content` works and does a reasonable thing, then `max-content` should be allowed, since fit-content is just "max-content, clamped between min-content and available space". So it can resolve to the same value as max-content, and relies on max-content *logic* to do so. Or, if there *is* a good reason to disallow `max-content`, then that same reason blocks `fit-content` (and its variants) as well.

I suspect we might indeed just want to hinge the logic on "not `auto`".

**Comment by frivoal** [comment]
FYI, on my computer (which is running macOS), ✅ looks identical to ✅︎, and ❌ looks identical to ❌︎, so the table in the original comment is not easily readable.

**Comment by Loirooriol** [comment]
Sorry, this is how it looks for me:

![](https://github.com/user-attachments/assets/a8b65187-d00d-4b5d-9efe-940d621e5d57)

**Comment by Loirooriol** [comment]
BTW, we didn't discuss this case in detail. @bfgeek interpreted the resolution as implying that `calc-size(auto, size)` allows fixed layout because it's not exactly `auto`.

But https://drafts.csswg.org/css-values-5/#resolving-calc-size

> A [calc-size()](https://drafts.csswg.org/css-values-5/#funcdef-calc-size) is treated, in all respects, as if it were its [calc-size basis](https://drafts.csswg.org/css-values-5/#calc-size-basis)

So not sure if `calc-size(auto, size)` should prevent fixed layout just like `auto`?

**Comment by bfgeek** [comment]
> BTW, we didn't discuss this case in detail. @bfgeek interpreted the resolution as implying that calc-size(auto, size) allows fixed layout because it's not exactly auto.

So I did this as it matches the "not initial" we've done for other cases in table layout.

**Comment by Loirooriol** [comment]
Can you provide some examples of these cases?

**Comment by bfgeek** [comment]
1. https://quirks.spec.whatwg.org/#the-table-cell-width-calculation-quirk
2. https://drafts.csswg.org/css-tables-3/#computing-the-table-height
3. https://drafts.csswg.org/css-tables-3/#used-width-of-table

**Comment by Loirooriol** [comment]
OK it probably makes more sense to discuss these things in another issue.

**Comment by bfgeek** [comment]
The resolution above isn't web-compatible (primarily due to ant-design - a widely used ui library - having it by default in a component).

WebKit also looks like the reverted for a similar reason. Likely will have to keep the quirk is `auto` && is `max-content`.

Ian

**Comment by tabatkins** [comment]
So just verifying - you're saying the resolution needs to be changed so that "auto" table layout is triggered by an `auto` *or* `max-content` width, and "fixed" layout is triggered for everything else?

**Comment by bfgeek** [comment]
> So just verifying - you're saying the resolution needs to be changed so that "auto" table layout is triggered by an auto or max-content width, and "fixed" layout is triggered for everything else?

Yup thats right.

**Comment by Loirooriol** [comment]
> So I did this as it matches the "not initial" we've done for other cases in table layout.

@bfgeek Now that it's no longer "not initial", do you think `calc-size()` should behave as its base?
